### PR TITLE
Feature/aggregation

### PR DESCRIPTION
**Pull request by jpbelanger-mtl**
Used base aggregation feature from https://github.com/Yelp/elastalert/pull/657 but modified it to be more flexible (generic querying and aggregation).

Added a daily alert feature to restrict the rule from running once it triggered.

### CODE REVIEW

File: elastalert/elastalert.py [hunk 1]
```diff
@@ -334,11 +348,72 @@ def get_hits(self, rule, starttime, endtime, index, scroll=False):
...
+        query = self.get_aggregation_query(base_query, rule['aggs']['aggs_form'])
+        elastalert_logger.info("Aggregation query: %s", query)
+
+        try:
+            res = self.current_es.search(index=index, body=query)
+            logging.debug(str(res))
+        except ElasticsearchException as e:
+            # Elasticsearch sometimes gives us GIGANTIC error messages
+            # (so big that they will fill the entire terminal buffer)
+            if len(str(e)) > 1024:
```
Comment (ndeslandes) [inline elastalert/elastalert.py]: lol

File: elastalert/elastalert.py [hunk 2]
```diff
@@ -334,11 +348,72 @@ def get_hits(self, rule, starttime, endtime, index, scroll=False):
...
+        return {'end_time': endtime, 'agg_num': aggregation_num, 'doc_count': hits, rule['timestamp_field']: ts_now()}
+
+    def get_query_daily_sent(self, rule, timestamp_field='@timestamp', to_ts_func=dt_to_ts):
+        start = ts_to_startday(ts_now())
+        end = start + datetime.timedelta(1)
+        starttime = to_ts_func(start)
+        endtime = to_ts_func(end)
+        filters = [{'query': {'match': {'alert_sent': {'query': 'true', 'type': 'phrase'}}}}, {'query': {'match': {'rule_name': {'query': rule['name'], 'type': 'phrase'}}}}]
+        es_filters = {'filter': {'bool': {'must': filters}}}
+        es_filters['filter']['bool']['must'].insert(0, {'range': {timestamp_field: {'gt': starttime,
```
Comment (ndeslandes) [inline elastalert/elastalert.py]: why not create the struct in one shot? or even create an object

Comment (jpbelanger-mtl) [inline elastalert/elastalert.py]: done!... much more readable too.

**Comment by ndeslandes** [comment]
LGTM

**Pull request merged**